src/lib/io/sol-aio-linux.c

```diff
@@ -0,0 +1,154 @@
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "aio");
+
+#include "sol-aio.h"
+
+#define AIO_BASE_PATH "/sys/bus/iio/devices"
+
+#define AIO_PATH(dst, device, pin) \
+    snprintf(dst, sizeof(dst), AIO_BASE_PATH "/iio:device%d/in_voltage%d_raw", device, pin)
+
+#define AIO_DEV_PATH(dst, device) \
+    snprintf(dst, sizeof(dst), AIO_BASE_PATH "/iio:device%d", device)
+
+struct sol_aio {
+    FILE *fp;
+    int device;
+    int pin;
+    unsigned int mask;
+};
+
+static bool
+_aio_open_fd(struct sol_aio *aio)
+{
+    int len;
+    char path[PATH_MAX];
+
+    len = AIO_PATH(path, aio->device, aio->pin);
+    if (len < 0 || len > PATH_MAX)
+        return false;
+
+    aio->fp = fopen(path, "re");
+    if (!aio->fp)
+        return false;
+    setvbuf(aio->fp, NULL, _IONBF, 0);
+
+    return true;
+}
+
+struct sol_aio *
+sol_aio_open_raw(const int device, const int pin, const unsigned int precision)
+{
+    int len;
+    char path[PATH_MAX];
+    struct stat st;
+    struct sol_aio *aio;
+
+    SOL_LOG_INTERNAL_INIT_ONCE;
+
+    if (!precision) {
+        SOL_WRN("aio #%d,%d: Invalid precision value=%d. Precision needs to be different of zero.",
+            device, pin, precision);
+        return NULL;
+    }
+
+    len = AIO_DEV_PATH(path, device);
+    if (len < 0 || len > PATH_MAX || stat(path, &st)) {
+        SOL_WRN("aio #%d,%d: aio device %d does not exist", device, pin, device);
+        return NULL;
+    }
+
+    len = AIO_PATH(path, device, pin);
+    if (len < 0 || len > PATH_MAX || stat(path, &st)) {
```

Comment:
It may be a good Idea to change this for the export part of gpio as well. Today we first use stat to see if its exported and open after that. We should try to open, try to export if fails and try open again. Is that right?